### PR TITLE
#221 Add percentileAll and medianAll

### PR DESCRIPTION
**Pull request by JingHuaMan**
#221 

Unlike `percentileBy` and `medianBy`, the function `percentileAllBy` and `medianAllBy` will retrieve all the values of the specific percentile.

The order of the output elements is the same as the order of those elements in the input `Seq`.

I add a lot of test cases for the two functions, and all the previous and new tests are passed.

Though there is already unsubmitted commits related to issue #221  in May 2020, it seems that there were some bugs with his codes.

### CODE REVIEW

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java [hunk 1]
```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
     }
 
     /**
-     * Get a {@link Collector} that calculates the common prefix of a set of strings.
+     * CS304 Issue link: https://github.com/jOOQ/jOOL/issues/221
```
Comment (lukaseder) [inline jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:944]: May I ask what CS304 means? I'll remove it again, but I'm curious. Is this some CS=Computer Science university class?

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java [hunk 2]
```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
...
+                        else
+                            right = mid;
+                    }
+
+                    for (; left < size && comparator.compare(l.get(left).v2, value) == 0; left++)
+                        result.add(l.get(left).v1);
+                    return Seq.seq(result);
+                }
+        );
+    }
```
Comment (lukaseder) [inline jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1019]: Have you checked, is there any refactoring potential to have common logic between `percentileBy()` and `percentileAllBy()`?

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java [hunk 3]
```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
...
+
+        if (percentile == 0.0)
+            // If percentile is 0, this is the same as taking the items with the minimum value.
+            return minAllBy(function, comparator);
+        else if (percentile == 1.0)
+            // If percentile is 1, this is the same as taking the items with the maximum value.
+            return maxAllBy(function, comparator);
+
+        return Collector.of(
+                () -> new ArrayList<Tuple2<? extends T, U>>(),
```
Comment (lukaseder) [inline jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:983]: Why is this covariance needed here?

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 2 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1019]: I mean, at least the collection to the accumulator seems to be the same, the difference being the finisher implementation.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java [hunk 4]
```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
...
+
+                    l.sort(Comparator.comparing(t -> t.v2, comparator));
+
+                    int left = 0;
+                    int right = (int) -Math.round(-(size * percentile + 0.5)) - 1;
+                    int mid;
+
+                    List<T> result = new ArrayList<>();
+                    U value = l.get(right).v2;
+                    // Search the first value equal to the value at the position of PERCENTILE by binary search
```
Comment (lukaseder) [inline jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: Nice thinking!

Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: Perhaps use `Collections.binarySearch()` instead, though?

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 1 above)
Comment (JingHuaMan) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:944]: I think we should apologize for these comments. We add these PRs because it is our software engineering course homework to make contributions on some open project╮(╯﹏╰）╭

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 2 above)
Comment (JingHuaMan) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1019]: I don't like the similar structure of `percentileBy` and `percentileAllBy` too, but I didn't find a proper way to refactor these codes.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 4 above)
Comment (JingHuaMan) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: I didn't know this API before. It's obviously better to use Java standard library.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 1 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:944]: No worries, just curious. I'll remove the comments. Another question, what made you choose jOOλ?

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 4 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: OK, I'll refactor it

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 2 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1019]: Sure, no worries. I'll refactor it. Thanks for the effort!

File: jOOL-java-8/src/test/java/org/jooq/lambda/CollectorTests.java [hunk 5]
```diff
@@ -308,7 +297,225 @@ public void testPercentileWithStringsAndFunction() {
...
+
+            @Override
+            public int compareTo(Item o) {
+                return this.val - o.val;
+            }
+
+            Item reverse() {
+                return new Item(-this.val);
+            }
+        }
```
Comment (lukaseder) [inline jOOL-java-8/src/test/java/org/jooq/lambda/CollectorTests.java:508]: I'll merge this, but I'll probably get rid of this class. I'm not sure what value it provides compared to just using `Integer`

**Comment by lukaseder** [comment]
Thanks again for your contribution. Note that if you plan on making more contributions, it will be sufficient to provide an implementation for the `jool` module only, the `jool-java-8` module is created automatically using scripts that aren't on this github repository.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 2 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1019]: This should do the trick:

```java
private static <T, U, R> Collector<T, List<Tuple2<T, U>>, R> percentileCollector(
    Function<? super T, ? extends U> function,
    Comparator<? super U> comparator,
    Supplier<R> onEmpty,
    Function<T, R> onSingle,
    Function<List<Tuple2<T, U>>, R> finisher
) {
    return Collector.of(
        () -> new ArrayList<>(),
        (l, v) -> l.add(tuple(v, function.apply(v))),
        (l1, l2) -> {
            l1.addAll(l2);
            return l1;
        },
        l -> {
            int size = l.size();

            if (size == 0)
                return onEmpty.get();
            else if (size == 1)
                return onSingle.apply(l.get(0).v1);

            l.sort(Comparator.comparing(t -> t.v2, comparator));
            return finisher.apply(l);
        }
    );
}
```

Leading to e.g.

```java
public static <T, U> Collector<T, ?, Optional<T>> percentileBy(double percentile, Function<? super T, ? extends U> function, Comparator<? super U> comparator) {
    if (percentile < 0.0 || percentile > 1.0)
        throw new IllegalArgumentException("Percentile must be between 0.0 and 1.0");

    if (percentile == 0.0)
        return minBy(function, comparator);
    else if (percentile == 1.0)
        return collectingAndThen(maxAllBy(function, comparator), s -> s.findLast());
    else
        return percentileCollector(
            function,
            comparator,
            Optional::empty,
            Optional::of,
            l -> Optional.of(l.get(percentileIndex(percentile, l.size())).v1)
        );
}
```

Where 

```java
private static int percentileIndex(double percentile, int size) {

    // x.5 should be rounded down
    return (int) -Math.round(-(size * percentile + 0.5)) - 1;
}
```

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 4 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: The JDK method finds a random index if multiple values have the same index, so we can't use that.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java [hunk 6]
```diff
@@ -941,8 +941,118 @@ else if (percentile == 1.0)
...
+                        mid = left + (right - left) / 2;
+                        if (comparator.compare(l.get(mid).v2, value) < 0)
+                            left = mid + 1;
+                        else
+                            right = mid;
+                    }
+
+                    for (; left < size && comparator.compare(l.get(left).v2, value) == 0; left++)
+                        result.add(l.get(left).v1);
+                    return Seq.seq(result);
```
Comment (lukaseder) [inline jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1016]: The auxiliary `result` list is not necessary. We could use `l` again here and add the comparator predicate as a `Seq.filter()`. Not sure though if we want to keep a reference to the possibly large array list, though... Edge cases...

File: jOOL-java-8/src/test/java/org/jooq/lambda/CollectorTests.java (hunk 5 above)
Comment (lukaseder) [reply jOOL-java-8/src/test/java/org/jooq/lambda/CollectorTests.java:508]: Nevermind. I see the point now. I'll just move the local class out of the method for better reuse.

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 4 above)
Comment (JingHuaMan) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:1005]: Yes, `Collections.binarySearch` will not return the first element equal to the target item. Part of the source codes of `Collections.binarySearch` attached.

```
while (low <= high) {
      int mid = (low + high) >>> 1;
      Comparable<? super T> midVal = get(i, mid);
      int cmp = midVal.compareTo(key);

      if (cmp < 0)
          low = mid + 1;
      else if (cmp > 0)
            high = mid - 1;
      else
            return mid; // key found
}
```

**Comment by lukaseder** [comment]
Additional changes here: https://github.com/jOOQ/jOOL/commit/d41bd549ebad07efc1010b841e037a493796c689

**Comment by lukaseder** [comment]
And here: https://github.com/jOOQ/jOOL/commit/1bdd69b2fd6f24fe193fbb53076cad0f33123108

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 1 above)
Comment (JingHuaMan) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:944]: It's mainly because this project is easy to learn, and it seems to be an important project because it's popular. Actually we have also considered JOOQ, but it's too big.

**Comment by lukaseder** [comment]
PS: Feel free to add yourself as `@author` to the `Agg` class

File: jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java (hunk 1 above)
Comment (lukaseder) [reply jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java:944]: Makes sense. Yes, it's much harder to make meaningful contributions to jOOQ - especially because the tests are not open source.
